sw/device/lib/base/hardened.h

```diff
@@ -32,4 +36,76 @@ typedef enum hardened_bool {
   kHardenedBoolFalse = 0x1d4u,
 } hardened_bool_t;
 
+/**
+ * Launders the 32-bit value `val`.
+ *
+ * `launder32()` is a pessimization hint that destroys any information the
+ * compiler had deduced about `val` at this point in the program. In other
+ * words, it behaves as if `val` was just read out of a `volatile` location for
+ * the purposes of global value numbering and strength reduction optimizations.
+ *
+ * However, this function does not have any reordering guarantees beyond those
+ * provided by the compiler's dependency analysis. All operations, including
+ * other launder operations and `volatile` operations, may be reordered past a
+ * call to `launder32()`.
+ *
+ * For example, in the following code, LLVM would fold `CHECK()` away, since a
+ * GVN analysis could tell it that the condition in the `CHECK()` is guaranteed
+ * to be `true`:
+ * ```
+ * if (x == 0) {
+ *   CHECK(x == 0);
+ * }
+ * ```
+ *
+ * On the other hand, if we believe that an attacker can glitch the chip in a
+ * way that causes the first branch to be skipped, this assumption in LLVM is no
+ * longer true. We can use laundering to ensure that the CHECK is not deleted,
+ * by preventing the compiler from identifying `x` as the same value in both
+ * expressions.
+ * ```
+ * if (launder32(x) == 0) {
+ *   CHECK(x == 0);
+ * }
+ * ```
+ *
+ * Note that the FIRST occurrence of `x` must be laundered for this to work. If
+ * we had written the `launder32()` in the second, the compiler would deduce
+ * that everywhere it is used, `x` can be constant-folded to zero. Although the
+ * branch would still be emitted (since LLVM cannot assume that `x == 0` implies
+ * that `launder32(x) == 0`), it can schedule instructions such that a constant
+ * zero ends up in `x`'s register. This snippet demonstrates the subtlety of
+ * this example in detail: https://godbolt.org/z/nc1bjaY87
```

Comment:
Is this link guaranteed to stay valid? It seems a bit odd to me to be linking out to a Godbolt page in checked-in code. Maybe this example (and others below) belong in local documentation?

Reply:
Godbolt's privacy policy seems to indicate this is not the case. I'm not deeply excited about pasting entire code snippets here inline, but I can just have a long-running giant example in the comments if we care.

Reply:
Well, that's up to the SW team: it was just a heads-up. Note that I suggested a document, not a giant comment. If you (plural) are happy with the external dependency, cool!

Reply:
I thought about it and instead expanded the comment to include more examples. Thanks for pointing this out, it wound up being useful.